python/dgl/dataloading/dataloader.py

```diff
@@ -1197,17 +1197,18 @@ def __init__(self, dataset, collate_fn=None, use_ddp=False, ddp_seed=0, **kwargs
             else:
                 dataloader_kwargs[k] = v
 
-        if collate_fn is None:
-            self.collate = GraphCollator(**collator_kwargs).collate
-        else:
-            self.collate = collate_fn
-
         self.use_ddp = use_ddp
         if use_ddp:
             self.dist_sampler = _create_dist_sampler(dataset, dataloader_kwargs, ddp_seed)
             dataloader_kwargs['sampler'] = self.dist_sampler
 
-        super().__init__(dataset=dataset, collate_fn=self.collate, **dataloader_kwargs)
+        super().__init__(dataset=dataset, collate_fn=collate_fn, **dataloader_kwargs)
+
+        # Overwrite PyTorch's default collator function with ours.
+        # After calling super().__init__() PyTorch should have set the collate_fn attribute.
+        # pylint: disable=access-member-before-definition
+        if self.collate_fn is torch.utils.data._utils.collate.default_collate:
+            self.collate_fn = GraphCollator(**collator_kwargs).collate
```

Comment:
I'm concerned we're relying on implementation details of torch's DataLoader, both in how it chooses to set its `self.collate_fn`, and that it make no other decisions in the constructor based on this. Could we instead explicitly check if batch_size is None, and set our collate_fn input to Torch's DataLoader based on that?
```python3
if batch_size is not None and collate_fn is None:
    collate_fn = GraphCollator(**collator_kwargs).collate
```
Or is there another reason we need to check what collator the parent DataLoader chooses?

Reply:
Makes sense.